File: tests/test_answer_guidance.py
import json
import unittest
from pathlib import Path
from agent import BatteryResearchAgent

ROOT = Path(__file__).resolve().parents[1]

class AnswerGuidanceTests(unittest.TestCase):
    def test_shared_answer_regression(self):
        agent = BatteryResearchAgent(fulltext_path=None, extra_data_path=None)
        cases = json.loads((ROOT / 'data/answer_regression.json').read_text(encoding='utf-8'))['cases']
        for case in cases:
            with self.subTest(question=case['question']):
                result = agent.answer(case['question'], use_rag=False)
                self.assertEqual(result['guidance_id'], case['guidance'])
                if 'task' in case:
                    self.assertEqual(result['task'], case['task'])
                for word in case['contains']:
                    self.assertIn(word, result['answer_markdown'])
                for word in case.get('excludes', []):
                    self.assertNotIn(word, result['answer_markdown'])

    def test_new_papers_exact_doi_lookup(self):
        agent = BatteryResearchAgent(fulltext_path=None, extra_data_path=None)
        for doi, paper_id in [('10.1038/s41586-018-0337-2', 'butler2018_ml_materials_review'), ('10.1038/s41524-026-02271-y', 'takenaka2026_sei_mlff')]:
            self.assertEqual(agent.search(doi, limit=1)[0]['id'], paper_id)

File: tests/test_answer_guidance.mjs
import fs from 'node:fs';
import vm from 'node:vm';
import assert from 'node:assert/strict';
const read = path => JSON.parse(fs.readFileSync(new URL('../'+path, import.meta.url),'utf8'));
const context = vm.createContext({window:{}});
vm.runInContext(fs.readFileSync(new URL('../static/browser_agent.js',import.meta.url),'utf8'), context);
const agent = context.window.BatteryBrowserAgent;
const papers = read('data/papers.json');
const config = read('data/search_config.json');
const cases = read('data/answer_regression.json').cases;
for (const c of cases) {
  const result = agent.answer(c.question,papers,config);
  assert.equal(result.guidance_id,c.guidance,c.question);
  if(c.task) assert.equal(result.task,c.task,c.question);
  for(const text of c.contains) assert.ok(result.answer_markdown.includes(text),`${c.question}: ${text}`);
  for(const text of c.excludes || []) assert.ok(!result.answer_markdown.includes(text),`${c.question}: unexpected ${text}`);
}
vm.runInContext('window.batteryBeginnerLessons={};',context);
vm.runInContext(fs.readFileSync(new URL('../static/md_lessons.js',import.meta.url),'utf8'),context);
const lessons=context.window.batteryMDLessons(()=>'<svg></svg>');
assert.equal(lessons.length,6);
assert.equal(context.window.batteryBeginnerLessons.md.length,lessons.length);
for(const [i,lesson] of lessons.entries()) {
  assert.ok(lesson.title && lesson.body && lesson.check);
  assert.equal(context.window.batteryBeginnerLessons.md[i].length,6);
}
console.log(`${cases.length} browser answer cases + 6 MD lesson contracts passed`);
